qiskit/transpiler/passes/scheduling/alap.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""ALAP Scheduling."""
+
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.passes.scheduling.asap import ASAPSchedule
+
+
+class ALAPSchedule(TransformationPass):
+    """ALAP Scheduling."""
+
+    def __init__(self, durations):
+        """ALAPSchedule initializer.
+
+        Args:
+            durations (InstructionDurations): Durations of instructions to be used in scheduling
+        """
+        super().__init__()
+        self._asap = ASAPSchedule(durations)
+
+    def run(self, dag):
+        """Run the ALAPSchedule pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to schedule.
+
+        Returns:
+            DAGCircuit: A scheduled DAG.
+
+        Raises:
+            TranspilerError: if the circuit is not mapped on physical qubits.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get('q', None) is None:
+            raise TranspilerError('ALAP schedule runs on physical circuits only')
+
+        new_dag = dag.reverse_ops()
+        new_dag = self._asap.run(new_dag)
+        new_dag = new_dag.reverse_ops()
```

Comment:
The DAG implementation of `reverse_ops` goes back-and-forth to circuit. So instead of doing that maybe just do so here so you don't add new things to DAG.